driver/src/driver/scheduler/system.rs

```diff
@@ -62,42 +65,27 @@ impl<'a> SystemScheduler<'a> {
         }
     }
 
-    /// Returns how long to sleep before starting the next iteration.
-    fn run_once(&mut self, now: SystemTime) -> Duration {
-        const RETRY_TIMEOUT: Duration = Duration::from_secs(10);
-        match self.determine_action(now) {
-            Ok(Action::Sleep(duration)) => {
-                info!("Sleeping {}s.", duration.as_secs());
-                duration
-            }
-            Ok(Action::Solve(batch_id, time_limit)) => {
-                info!(
-                    "Starting to solve batch {} with time limit {}s.",
-                    batch_id.0,
-                    time_limit.as_secs_f64()
-                );
-                match self.driver.run(batch_id.0.into(), time_limit) {
-                    DriverResult::Ok => {
-                        info!("Batch {} solved successfully.", batch_id.0);
-                        self.last_solved_batch.replace(batch_id);
-                        Duration::from_secs(0)
-                    }
-                    DriverResult::Retry(err) => {
-                        error!("StableXDriver retryable error: {}", err);
-                        RETRY_TIMEOUT
-                    }
-                    DriverResult::Skip(err) => {
-                        error!("StableXDriver unretryable error: {}", err);
-                        self.last_solved_batch.replace(batch_id);
-                        Duration::from_secs(0)
-                    }
-                }
+    fn start_solving_in_thread<'b>(
+        &self,
+        batch_id: BatchId,
+        solver_time_limit: Instant,
+        scope: &Scope<'b>,
+    ) where
+        'a: 'b,
+    {
+        let driver = self.driver;
+        scope.spawn(move |_| loop {
```

Comment:
```suggestion
        scope.spawn(move |_| while let Some(time_limit) = solver_time_limit.checked_duration_since(Instant::now()) {
```

That way we have a clear exit condition in the loop.